File: haskell/Yacht.hs
module Yacht where

import Data.List (sort)

data Category = Ones
              | Twos
              | Threes
              | Fours
              | Fives
              | Sixes
              | FullHouse
              | FourOfAKind
              | LittleStraight
              | BigStraight
              | Choice
              | Yacht

yacht :: Category -> [Int] -> Int
yacht Yacht list          = scoreYacht list
yacht Choice list         = sum list
yacht BigStraight list    = scoreBigStraight $ sort list
yacht LittleStraight list = scoreLittleSt8 $ sort list
yacht FullHouse list      = scoreFullHouse $ sort list
yacht FourOfAKind list    = scoreFourOfAKind $ sort list
yacht x list              = scoreSingles (categoryToInt x) list

scoreYacht :: [Int] -> Int
scoreYacht (a:b:c:d:e:_) =
  if (a == b && b == c && c == d && d == e)
    then 50
    else 0

scoreBigStraight :: [Int] -> Int
scoreBigStraight (2:3:4:5:6:_) = 30
scoreBigStraight _ = 0

scoreLittleSt8 :: [Int] -> Int
scoreLittleSt8 (1:2:3:4:5:_) = 30
scoreLittleSt8 _ = 0

scoreFourOfAKind :: [Int] -> Int
scoreFourOfAKind (a:b:c:d:e:_) =
  if (a == b && b == c && c == d) || (b == c && c == d && d == c)
    then 4 * b
    else 0

scoreFullHouse :: [Int] -> Int
scoreFullHouse (a:b:c:d:e:_) =
  if (a == b && b == c && d == e && c /= d) ||
    (a == b && c == d && d == e && b /= c)
    then a + b + c + d + e
    else 0

categoryToInt :: Category -> Int
categoryToInt Ones = 1
categoryToInt Twos = 2
categoryToInt Threes = 3
categoryToInt Fours = 4
categoryToInt Fives = 5
categoryToInt Sixes = 6
categoryToInt _ = 0

scoreSingles :: Int -> [Int] -> Int
scoreSingles n list = sum $ filter (== n) list

File: elixir/yacht.ex
defmodule Yacht do
  @type category ::
          :ones
          | :twos
          | :threes
          | :fours
          | :fives
          | :sixes
          | :full_house
          | :four_of_a_kind
          | :little_straight
          | :big_straight
          | :choice
          | :yacht

  @score_multipliers %{
    ones: 1,
    twos: 2,
    threes: 3,
    fours: 4,
    fives: 5,
    sixes: 6
  }

  @doc """
  Calculate the score of 5 dice using the given category's scoring method.
  """
  @spec score(category :: category(), dice :: [integer]) :: integer
  def score(category, dice) when category in [:ones, :twos, :threes, :fours, :fives, :sixes] do
    dice
    |> Enum.count(&(&1 == @score_multipliers[category]))
    |> Kernel.*(@score_multipliers[category])
  end

  def score(:yacht, [x, x, x, x, x]), do: 50
  def score(:yacht, _), do: 0

  def score(:full_house, dice) do
    dice
    |> Enum.sort()
    |> score_fullhouse()
  end

  def score(:four_of_a_kind, dice) do
    dice
    |> Enum.sort()
    |> score_four_of_a_kind()
  end

  def score(:little_straight, dice) do
    if Enum.sort(dice) == [1, 2, 3, 4, 5] do
      30
    else
      0
    end
  end

  def score(:big_straight, dice) do
    if Enum.sort(dice) == [2, 3, 4, 5, 6] do
      30
    else
      0
    end
  end

  def score(:choice, dice), do: Enum.sum(dice)

  defp score_fullhouse([x, x, x, y, y]) when x != y, do: 3 * x + 2 * y
  defp score_fullhouse([x, x, y, y, y]) when x != y, do: 2 * x + 3 * y
  defp score_fullhouse(_), do: 0

  defp score_four_of_a_kind([x, x, x, x, _]), do: 4 * x
  defp score_four_of_a_kind([_, x, x, x, x]), do: 4 * x
  defp score_four_of_a_kind(_), do: 0
end
